libcxx/include/__type_traits/desugars_to.h

```diff
@@ -52,6 +52,24 @@ struct __totally_ordered_less_tag {};
 template <class _CanonicalTag, class _Operation, class... _Args>
 inline const bool __desugars_to_v = false;
 
+// For the purpose of determining whether something desugars to something else,
```

Comment:
Do we have tests where we test this desugaring  with cvref types in our algorithms?
```
// syntactically, the operation is equivalent to calling `a < b`, and these expressions
// have to be true for any `a` and `b`:
// - `(a < b) == (b > a)`
```
Here if `a` and `b` are a `volatile ref` to the same object `c` the statement above may not hold true and can change every time it's tested.

My concern is specifically about `volatile`.

Reply:
There is a subtlety here. This patch only disregards the cv-ref qualifiers on the operation itself, which would be `std::less` in this case. It does not disregard the cv-ref qualifiers on the arguments that are passed to the function.

Reply:
Still I'm concerned that a `volatile` qualified operator is intended not to be optimized. I expect them to be very rare so not desugaring them sounds not a big loss. IMO if we want to desugar them I'd like some tests that desugaring them has no unexpected issues in cases where a `volatile` operator depends on read and writes not being optimized away.